beacon-chain/utils/shuffle.go

```diff
@@ -161,3 +142,140 @@ func PermutedIndex(index uint64, listSize uint64, seed [32]byte) (uint64, error)
 func SplitOffset(listSize uint64, chunks uint64, index uint64) uint64 {
 	return (listSize * index) / chunks
 }
+
+// ShuffleList returns list of shuffled indexes in a pseudorandom permutation `p` of `0...list_size - 1` with ``seed`` as entropy.
+// We utilize 'swap or not' shuffling in this implementation; we are allocating the memory with the seed that stays
+// constant between iterations instead of reallocating it each iteration as in the spec. This implementation is based
+// on the original implementation from protolambda, https://github.com/protolambda/eth2-shuffle
+func ShuffleList(input []uint64, seed [32]byte) ([]uint64, error) {
+	return innerShuffleList(input, seed, true)
+}
+
+// UnshuffleList Un-shuffles the list by running backwards through the round count.
+func UnshuffleList(input []uint64, seed [32]byte) ([]uint64, error) {
+	return innerShuffleList(input, seed, false)
+}
+
+// shuffles or unshuffles, depending on the `dir` true for shuffling, false for unshuffling.
+func innerShuffleList(input []uint64, seed [32]byte, dir bool) ([]uint64, error) {
+	if len(input) <= 1 {
+		// nothing to (un)shuffle
+		return input, nil
+	}
+	if uint64(len(input)) > maxShuffleListSize {
+		return nil, fmt.Errorf("list size %d out of bounds",
+			len(input))
+	}
+	rounds := uint8(params.BeaconConfig().ShuffleRoundCount)
+	if rounds == 0 {
+		return input, nil
+	}
+
+	listSize := uint64(len(input))
+	buf := make([]byte, totalSize, totalSize)
+	r := uint8(0)
+	if !dir {
+		// Start at last round.
+		// Iterating through the rounds in reverse, un-swaps everything, effectively un-shuffling the list.
+		r = rounds - 1
+	}
+	// Seed is always the first 32 bytes of the hash input, we never have to change this part of the buffer.
+	copy(buf[:seedSize], seed[:])
+	for {
+		// spec: pivot = bytes_to_int(hash(seed + int_to_bytes1(round))[0:8]) % list_size
```

Comment:
Please move all of this inline commentary to the godoc comment. It is too noisy and doesn't helpful readability.

Reply:
moved